### PR TITLE
Fix attempt for Google and "All day".

### PR DESCRIPTION
**Pull request by vever001**
## Context and Purposes

Follow up of https://github.com/spatie/calendar-links/pull/184 and comment https://github.com/spatie/calendar-links/pull/184#discussion_r1470231149

### CODE REVIEW

File: src/Generators/Google.php [hunk 1]
```diff
@@ -14,17 +14,15 @@ class Google implements Generator
     /** @var string {@see https://www.php.net/manual/en/function.date.php} */
     protected $dateFormat = 'Ymd';
     /** @var string */
-    protected $dateTimeFormat = 'Ymd\THis\Z';
+    protected $dateTimeFormat = 'Ymd\THis';
```
Comment (vever001) [inline src/Generators/Google.php:17]: Since we are passing the TZ below to Google, I think we shouldn't add the `Z`

File: src/Link.php [hunk 2]
```diff
@@ -64,10 +64,8 @@ public static function create(string $title, \DateTimeInterface $from, \DateTime
     {
         // When creating all day events, we need to be in the UTC timezone as all day events are "floating" based on the user's timezone
         if ($allDay) {
-            $startDate = new \DateTime($from->format('Y-m-d'), new \DateTimeZone('UTC'));
             $numberOfDays = $from->diff($to)->days + 1;
-
-            return self::createAllDay($title, $startDate, $numberOfDays);
+            return self::createAllDay($title, $from, $numberOfDays);
```
Comment (vever001) [inline src/Link.php:68]: Why was it dropping the time here? This results in wrong dates down the line I believe.

File: src/Link.php [hunk 3]
```diff
@@ -83,12 +81,7 @@ public static function create(string $title, \DateTimeInterface $from, \DateTime
      */
     public static function createAllDay(string $title, \DateTimeInterface $fromDate, int $numberOfDays = 1): self
     {
-        // In cases where the from date is not UTC, make sure it's UTC, size all day events are floating and non UTC dates cause bugs in the generators
-        if ($fromDate->getTimezone() !== new \DateTimeZone('UTC')) {
-            $fromDate = \DateTime::createFromFormat('Y-m-d', $fromDate->format('Y-m-d'));
-        }
```
Comment (vever001) [inline src/Link.php:89]: This condition doesn't seem towork as see here: https://3v4l.org/FAnI5

File: tests/TestCase.php [hunk 4]
```diff
@@ -63,7 +63,7 @@ protected function createMultipleDaysAllDayEventLink(bool $immutable = false): L
 
         return Link::createAllDay(
             'Birthday',
-            $dateTimeClass::createFromFormat('Y-m-d', '2018-02-01', new DateTimeZone('UTC')),
+            $dateTimeClass::createFromFormat('Y-m-d H:i', '2018-02-01 00:00', new DateTimeZone('UTC')),
```
Comment (vever001) [inline tests/TestCase.php:66]: Set fixed time to prevent random UIDs during tests.

File: src/Generators/Google.php [hunk 5]
```diff
@@ -21,19 +21,9 @@ public function generate(Link $link): string
...
-        // Add timezone name if it is specified in both from and to dates and is the same for both
-        if (
-            $link->from->getTimezone() && $link->to->getTimezone()
-            && $link->from->getTimezone()->getName() === $link->to->getTimezone()->getName()
-        ) {
-            $url .= '&ctz=' . $link->from->getTimezone()->getName();
-        }
-
+        $url .= '&dates='.$link->from->format($dateTimeFormat).'/'.$link->to->format($dateTimeFormat);
+        $url .= '&ctz=UTC';
```
Comment (vever001) [inline src/Generators/Google.php:26]: This formats dates in UTC (using `Ymd\THis\Z`) like the other generators do, and forces `ctz=UTC`

File: src/Generators/Google.php [hunk 6]
```diff
@@ -14,26 +14,16 @@ class Google implements Generator
...
         $dateTimeFormat = $link->allDay ? $this->dateFormat : $this->dateTimeFormat;
-        $url .= '&dates='.$utcStartDateTime->format($dateTimeFormat).'/'.$utcEndDateTime->format($dateTimeFormat);
-
-        // Add timezone name if it is specified in both from and to dates and is the same for both
-        if (
-            $link->from->getTimezone() && $link->to->getTimezone()
-            && $link->from->getTimezone()->getName() === $link->to->getTimezone()->getName()
-        ) {
-            $url .= '&ctz=' . $link->from->getTimezone()->getName();
-        }
```
Comment (vever001) [inline src/Generators/Google.php:35]: I moved this timezone match check to `Link::__construct`, for consistency.

File: src/Generators/Google.php [hunk 7]
```diff
@@ -14,26 +14,16 @@ class Google implements Generator
...
-
+        $url .= '&dates='.$link->from->format($dateTimeFormat).'/'.$link->to->format($dateTimeFormat);
+        $url .= '&ctz=' . $link->from->getTimezone()->getName();
```
Comment (vever001) [inline src/Generators/Google.php:26]: This keeps the original TZ and passes it to google.
Also, it seems the `\Z` in $dateTimeFormat causes issues with full dates with time and Google thinks dates are given in UTC.

File: tests/TestCase.php [hunk 8]
```diff
@@ -63,7 +63,7 @@ protected function createMultipleDaysAllDayEventLink(bool $immutable = false): L
 
         return Link::createAllDay(
             'Birthday',
-            $dateTimeClass::createFromFormat('Y-m-d', '2018-02-01', new DateTimeZone('UTC')),
+            $dateTimeClass::createFromFormat('Y-m-d', '2018-02-01', new DateTimeZone('UTC'))->setTime(0, 0),
```
Comment (vever001) [inline tests/TestCase.php:66]: Set fixed time to prevent random UIDs during tests.

File: tests/Generators/GoogleGeneratorTest.php [hunk 9]
```diff
@@ -35,4 +35,12 @@ public function it_correctly_generates_all_day_events_by_dates(): void
             $this->generator()->generate($this->createEventMultipleDaysViaStartEndWithTimezoneLink())
         );
     }
+
+    /** @test */
+    public function it_correctly_generates_all_day_events_by_dates_diff_tz(): void
```
Comment (alies-dev) [inline tests/Generators/GoogleGeneratorTest.php:40]: I think having this test is beneficial for all generators. How about moving it to the `GeneratorTestContract`?

**Review by alies-dev — COMMENTED** [review]
that's a great fix! Thank you a lot for your work on this, @vever001!

I have one minor note, please take a look.

File: src/Generators/Google.php (hunk 6 above)
Comment (atymic) [reply src/Generators/Google.php:45]: Good call!

File: src/Link.php [hunk 10]
```diff
@@ -86,12 +91,7 @@ public static function create(string $title, \DateTimeInterface $from, \DateTime
...
-
-        $from = (clone $fromDate)->modify('midnight');
+        $from = (clone $fromDate);
```
Comment (atymic) [inline src/Link.php:94]: With my old code, we're removing the hours/days portion. I guess this probably wasn't required and it only was done for non UTC dates.

File: tests/Generators/GoogleGeneratorTest.php (hunk 9 above)
Comment (atymic) [reply tests/Generators/GoogleGeneratorTest.php:41]: Agreed!
Maybe also moving the ones I added as well?

**Review by atymic — APPROVED** [review]
Awesome to see all the activity here in this repo :)
Good job all!

**Comment by atymic** [comment]
Failing test looks fine as well

**Comment by atymic** [comment]
Yea, it's a shame we can't test them again the provider itself.

**Comment by alies-dev** [comment]
Thanks @vever001 and @atymic 
I'm happy to see your activity here. I think the PR is almost done; even at this stage it's still an improvement. But it's still cab be improved. I'm ok to merge it as is and release and wait for a new PR to improve it even more. But I would also love if you make these improvements within this PR. That's up to you, I'm ok with both options, just want to keep the tempo and make this awesome job done :)

**Comment by samuelreichor** [comment]
Hi guys, I'm looking forward for the merge :)

**Comment by vever001** [comment]
Hi @alies-dev, sure please do :)
I just didn't have time yet to improve tests, comments https://github.com/spatie/calendar-links/pull/192#discussion_r1470632684 and https://github.com/spatie/calendar-links/pull/192#discussion_r1470714758
But any help is welcome.

--
I have some ideas on how to improve this lib but this would clearly be out of scope for this issue and introduce breaking changes. See for example this https://github.com/vever001/calendar-links/commit/39a9bbd22ceb0da9dc4f93bef6286324d9a48dba
Also some kind of plugin system would be nice (with class discovery, plugins using PHP 8 class attributes?) but I don't know if the requirement here are to not have any dependencies (e.g: symfony or other).

**Review by alies-dev — APPROVED** [review]
@vever001 
Great! It can be improved even more, but I believe  the PR in it's current shape is still an improvement and thus we can merge it! Thank you for your amazing work!

**Pull request merged**